### ISSUE TITLE
new lint similar to `or_fun_call` but only allowing eager evaluation of primitives

### ISSUE THREAD

**Issue by gner56** [rust-lang/rust-clippy#17279]
### What it does

# Overview
This lint should be similar to `or_fun_call`, maybe a name for it can be `or_non_primitives`. The idea it would disallow eager evaluation of values beyond obviously primitive cases, e.g. `ok_or(1)` or `ok_or(true)`

# Motivation
Certain things which are not calls (and hence not denied by `or_fun_call`), such as constructing certain enums or structs can be very expensive if eagerly evaluated. This can cause real performance issues/bottlenecks, especially in hot-paths where a seemingly cheap call (e.g. error construction) may become expensive if the underlying enum/struct is changed in the future.

# Example (not caught by or_fun_call)
```rust
#![deny(clippy::or_fun_call)]
#[derive(Debug)]
struct ExpensiveStruct {
    #[allow(dead_code)]
    message: String,
}

fn main() -> Result<(), ExpensiveEnum> {
    let maybe_value: Option<i32> = Some(42);
    let message = "x".repeat(10_000_000); // or String::from("some_long_string"), etc

    // This enum can be expensive to construct, even though it isn't a function call.
    let value = maybe_value.ok_or(ExpensiveStruct {
        message
    })?;

    println!("{value}");
    Ok(())
}
```

As an example, Clippy does not warn on the above code snippet, even though the enum/struct could be arbitrarily nested and/or could be expensive or slow to construct. 

I am a beginner to contributing here, and am willing to work on this as well - but just wanted to get feedback/discussion about the above. Thank you!
